Add support for checkpoint image

rst0git opened the pull request:
This is an enhancement proposal for the checkpoint / restore feature of Podman that enables container migration across multiple systems with standard image distribution infrastructure.

A new option `--create-image <image>` has been added to the `podman container checkpoint` command. This option tells Podman to create a container image.  This is a standard image with a single layer, tar archive, that that contains all checkpoint files. This is similar to the current approach for container migration using `--export` / `--import`.

The checkpoint image can be pushed to a container registry and pulled on a different system.  It can also be exported locally with `podman image save` and inspected with `podman inspect`. Inspecting the image would display additional information (stored as annotations) about the host and the versions of Podman, criu, crun/runc, kernel, etc.

`podman container restore` has also been extended to support checkpoint image provided as input.

Simple example:
```
podman run -d --name looper busybox /bin/sh -c 'i=0; while true; do echo $i; i=$(expr $i + 1); sleep 1; done'
podman container checkpoint --create-image checkpoint-image-test looper
podman rm looper
podman container restore checkpoint-image-test
```
Example with multiple containers:
```
podman run -d --name looper-1 busybox /bin/sh -c 'i=0; while true; do echo $i; i=$(expr $i + 1); sleep 1; done'
podman run -d --name looper-2 busybox /bin/sh -c 'i=0; while true; do echo $i; i=$(expr $i + 1); sleep 1; done'
podman container checkpoint --create-image checkpoint-1 looper-1
podman container checkpoint --create-image checkpoint-2 looper-2
podman rm looper-1 looper-2
podman container restore checkpoint-1 checkpoint-2
```
Example using container registry:
```
podman run -d --name looper busybox /bin/sh -c 'i=0; while true; do echo $i; i=$(expr $i + 1); sleep 1; done'
podman container checkpoint --create-image quay.io/rst0git/checkpoint-image-test-1 looper
podman push quay.io/rst0git/checkpoint-image-test-1
podman rm -a
podman rmi -a
podman pull quay.io/rst0git/checkpoint-image-test-1
podman inspect quay.io/rst0git/checkpoint-image-test-1
podman container restore quay.io/rst0git/checkpoint-image-test-1
```

vrothberg commented:
@adrianreber PTAL

adrianreber commented:
Overall I like this approach a lot. This is also very helpful when looking at our upcoming CRI-O changes. With something like this we might get container restore in Kubernetes without Kubernetes knowing that it is a restore. At some point, not now, this can be enhanced to have `podman run registry/image:latest` automatically do a restore if the image is a checkpoint image.

@rst0git Do you also store the name of the container engine (in this case Podman) in the image metadata? Just to make sure we have something in the image to quickly see if it is an image that we can also restore in CRI-O. I have already restored Podman checkpoints in CRI-O using the existing checkpoint archive. My assumption is that it should also be possible with this image format.

Luap99 commented on pkg/checkpoint/checkpoint_restore.go:
```diff
@@ -39,7 +37,12 @@ func CRImportCheckpoint(ctx context.Context, runtime *libpod.Runtime, restoreOpt
 
+// CRImportCheckpoint it the function which imports the information
+// from checkpoint tarball and re-creates the container from that information
+func CRImportCheckpoint(ctx context.Context, runtime *libpod.Runtime, restoreOptions entities.RestoreOptions, dir string) ([]*libpod.Container, error) {
```
This is a breaking change, you add a new parameter to this function.

vrothberg replied:
Why do we care about API stability in `pkg/checkpoint`?

vrothberg commented on utils/utils.go:
```diff
@@ -120,6 +122,42 @@ func CreateTarFromSrc(source string, dest string) error {
 	return TarToFilesystem(source, file)
 }
 
+func CopyDir(src string, dest string, excludeFiles []string) error {
```
Yes, it's probably a good idea using that since it's secure and well tested. But it's more complicated to use with overhead of (un)tar'ing things.

Luap99 replied:
Yea the secure part will be important given that this copies right from the image mount point. I imagine there can be problems with symlinks, etc...

rst0git replied:
Thanks Valentin and Paul, I've discussed this with @adrianreber and updated the pull request to use `CopyWithTar` from `archive` instead.

rst0git commented:
> Do you also store the name of the container engine (in this case Podman) in the image metadata? 

@adrianreber Not explicitly, but the annotation keys contain the name of the container engine (e.g., [`io.podman.annotations.`](https://github.com/containers/podman/blob/bbae81a291159895d1dba841a591293eec6607c2/libpod/define/annotations.go)).

rst0git commented on libpod/container_internal_linux.go:
```diff
@@ -1088,6 +1090,171 @@ func (c *Container) addNamespaceContainer(g *generate.Generator, ns LinuxNS, ctr
+	}
+	if imageExists {
+		return errors.Errorf("Image with this name already exists")
+	}
```
> Is this really an issue? The image with the name conflict would be untagged in this case.

No, I don't think so. The untagged image would contain all checkpoint annotations as well as `NamesHistory` annotation that includes the names previously associated with the image. I believe it is safe to remove this check.

rst0git commented on libpod/container_internal_linux.go:
```diff
@@ -1088,6 +1090,171 @@ func (c *Container) addNamespaceContainer(g *generate.Generator, ns LinuxNS, ctr
+	// Add image annotations with information about the container and the host.
+	// This information is useful to check compatibility before restoring the checkpoint
+	importBuilder.SetAnnotation(
+		"application/vnd.podman.container.checkpoint.name",
```
Thanks, this is a good idea!

rst0git commented:
I have updated the PR to address the code review comments and enable restore of checkpoint images with `podman-remote`.

TomSweeneyRedHat commented on docs/source/markdown/podman-container-checkpoint.1.md:
```diff
@@ -28,6 +28,58 @@ archives. Not compressing the checkpoint archive can result in faster checkpoint
+pushed to a standard container registry and pulled on a different system. It can
+also be exported locally with **podman image save** and inspected with **podman
+inspect**. Inspecting a checkpoint image would display additional information,
+stored as annotations, about the host environment used to the checkpoint:
```
```suggestion
stored as annotations, about the host environment used for the checkpoint:
```
or perhaps "to do the checkpoint:"

TomSweeneyRedHat commented on docs/source/markdown/podman-container-checkpoint.1.md:
```diff
@@ -28,6 +28,58 @@ archives. Not compressing the checkpoint archive can result in faster checkpoint
+stored as annotations, about the host environment used to the checkpoint:
+
+- **io.podman.annotations.checkpoint.name**: Human-readable name
+  of the original container.
```
container or container image?

rst0git replied:
This is the container name. The image name is stored in `rawImageName` and `rootfsImageName`.

This is useful, for example, when multiple containers with different names have been created from the same image.

TomSweeneyRedHat commented on docs/source/markdown/podman-container-checkpoint.1.md:
```diff
@@ -28,6 +28,58 @@ archives. Not compressing the checkpoint archive can result in faster checkpoint
+pushed to a standard container registry and pulled on a different system. It can
+also be exported locally with **podman image save** and inspected with **podman
+inspect**. Inspecting a checkpoint image would display additional information,
+stored as annotations, about the host environment used to the checkpoint:
```
You make several mentions of "container" below in the explanation of the various fields, but the work "container" doesn't appear once in the description, just image.  Maybe the first line should be "Create a checkpoint image from the provided container." or some such.

rst0git replied:
Thanks, I've updated this paragraph.

TomSweeneyRedHat commented on docs/source/markdown/podman-container-restore.1.md:
```diff
@@ -106,14 +107,16 @@ If the **--name, -n** option is used, Podman will not attempt to assign the same
 be used once and the restored *container* will have another IP address. This also means
 that **--name, -n** cannot be used in combination with **--tcp-established**.\
-*IMPORTANT: This OPTION is only available in combination with __--import, -i__.*
+*IMPORTANT: This OPTION is only available for checkpoint image or in combination
```
```suggestion
*IMPORTANT: This OPTION is only available for a checkpoint image or in combination
```

TomSweeneyRedHat commented on docs/source/markdown/podman-container-restore.1.md:
```diff
@@ -106,14 +107,16 @@ If the **--name, -n** option is used, Podman will not attempt to assign the same
 has to have the same namespaces shared as the pod this container was checkpointed
 from (see **[podman pod create --share](podman-pod-create.1.md#--share)**).\
-*IMPORTANT: This OPTION is only available in combination with __--import, -i__.*
+*IMPORTANT: This OPTION is only available for checkpoint image or in combination
```
```suggestion
*IMPORTANT: This OPTION is only available for a checkpoint image or in combination
```

TomSweeneyRedHat commented on docs/source/markdown/podman-container-restore.1.md:
````diff
@@ -175,6 +178,15 @@ $ podman run --rm -p 2345:80 -d webserver
+Start a container with name "foobar-1". Create a checkpoint image "foobar-checkpoint". Restore the container from checkpoint image with different name.
+```
+# podman run --name foobar-1 -d webserver
+# podman container checkpoint --create-image foobar-checkpoint foobar-1
````
Here you call it `foobar-checkpoint`, but below it's `foobar-checkpoint-image`.  Does `-image` get added automatically or is it just missing in this line?

rst0git replied:
Thanks for catching this!

rst0git commented:
@vrothberg @Luap99 @TomSweeneyRedHat
Thank you for your code review. I have updated the pull request based on your suggestions, rebased on the main branch, and added a few tests.

vrothberg commented on cmd/podman/containers/restore.go:
```diff
@@ -137,6 +165,9 @@ func restore(cmd *cobra.Command, args []string) error {
 	}
+	if argLen > 1 && restoreOptions.Name != "" {
+		return errors.Errorf("--name can only be used with one checkpoint image")
+	}
```
I think that all these checks should happen _before_ image inspect.  It works as is just fine but the code is easier to maintain when we validate input before using it.

rst0git replied:
This is slightly complicated because before image inspect we don't know if the input is an image or a container.
- If it is a container, `--name` is not supported.
- If it is an image, it should be only one.

vrothberg commented on cmd/podman/containers/restore.go:
```diff
@@ -137,6 +165,9 @@ func restore(cmd *cobra.Command, args []string) error {
 		return errors.Errorf("you must provide at least one name or id")
 	}
+	if argLen > 1 && restoreOptions.Name != "" {
+		return errors.Errorf("--name can only be used with one checkpoint image")
```
Minor nit (non-blocking): consider using `fmt.Errorf` and `errors.New` instead of the non-standard `errors` package.

vrothberg commented on docs/source/markdown/podman-container-checkpoint.1.md:
```diff
@@ -28,6 +28,60 @@ archives. Not compressing the checkpoint archive can result in faster checkpoint
+
+Create a checkpoint image from the provided container. A checkpoint image is a
+standard OCI image with single layer (tar file) that contains all checkpoint
+files. The content of the image is in the same format as tar.gz file created
```
This reads like the image would be stored as a tar.gz file.  But isn't the image created in the local image store?

rst0git replied:
Yes, you are right. The image is created in the local image store. I will try to rephrase this.

vrothberg commented on libpod/container_api.go:
```diff
@@ -781,6 +784,10 @@ type ContainerCheckpointOptions struct {
 	// images. One is TargetFile, the other is ImportPrevious.
 	ImportPrevious string
+	// CreateImage tells Podman to create an OCI image from container
+	// checkpoint. The image can be pushed to remote image registry,
+	// pulled and restored on a different node.
```
```suggestion
	// checkpoint in the local image store.
```

vrothberg commented on libpod/container_internal_linux.go:
```diff
@@ -1098,6 +1100,160 @@ func (c *Container) addNamespaceContainer(g *generate.Generator, ns LinuxNS, ctr
+	return nil
+}
+
+func (c *Container) validateCheckpointImageName(options *ContainerCheckpointOptions) error {
```
```suggestion
func (c *Container) resolveCheckpointImageName(options *ContainerCheckpointOptions) error {
```
Nit: no real validation is happening

vrothberg commented on cmd/podman/containers/restore.go:
```diff
@@ -97,20 +99,46 @@ func restore(cmd *cobra.Command, args []string) error {
-	if restoreOptions.Import == "" && restoreOptions.ImportPrevious != "" {
-		return errors.Errorf("--import-previous can only be used with --import")
+
+	// Find out if this is an image or container ID
```
The comment seems misleading to me as we only check if it's an image. At least I cannot spot a check if the input may refer to a container.

vrothberg reviewed (commented):
Thanks a lot for the great work!  I just have a some minor final nits.

Please make sure to squash the commits into a single one before merging.

rst0git commented:
> Please make sure to squash the commits into a single one before merging.

I would be happy to squash the commits but `Build Each Commit` fails when a single commit increases the size of `bin/podman` with more than 51200 bytes: https://cirrus-ci.com/task/6191166545723392?logs=main#L144

Luap99 commented:
> I would be happy to squash the commits but `Build Each Commit` fails when a single commit increases the size of `bin/podman` with more than 51200 bytes: https://cirrus-ci.com/task/6191166545723392?logs=main#L144

We can overwrite this with a label.

Luap99 commented on libpod/container_internal_linux.go:
```diff
@@ -1098,6 +1100,160 @@ func (c *Container) addNamespaceContainer(g *generate.Generator, ns LinuxNS, ctr
+	// This information is useful to check compatibility before restoring the checkpoint
+	importBuilder.SetAnnotation(
+		define.CheckpointAnnotationName,
+		c.Config().Name,
```
Do not call `c.Config()` just use `c.config`, `Config()` return a deep copy and thus is very slow.

Luap99 commented on libpod/container_internal_linux.go:
```diff
@@ -1098,6 +1100,160 @@ func (c *Container) addNamespaceContainer(g *generate.Generator, ns LinuxNS, ctr
+	)
+	importBuilder.SetAnnotation(
+		define.CheckpointAnnotationRawImageName,
+		c.Config().RawImageName,
```
same here

rst0git commented:
Thanks @Luap99, I squashed the changes into 3 commits.

Luap99 commented on libpod/container_internal_linux.go:
```diff
@@ -1098,6 +1100,160 @@ func (c *Container) addNamespaceContainer(g *generate.Generator, ns LinuxNS, ctr
+
+	// Add image annotations with information about the container and the host.
+	// This information is useful to check compatibility before restoring the checkpoint
+	importBuilder.SetAnnotation(
```
Can we turn this logic into a loop. Create a map[string]string will all annotations and then just loop over it.
I think that would look cleaner but I am also fine with this.

Luap99 commented on cmd/podman/containers/checkpoint.go:
```diff
@@ -68,6 +68,10 @@ func init() {
 
+	createImageFlagName := "create-image"
+	flags.StringVarP(&checkpointOptions.CreateImage, createImageFlagName, "", "", "Create checkpoint image with specified name")
+	_ = checkpointCommand.RegisterFlagCompletionFunc(createImageFlagName, completion.AutocompleteDefault)
```
AutocompleteDefault will complete paths. I don't think that makes sense for this option. It should be either images or none.

rst0git commented:
> > I would be happy to squash the commits but `Build Each Commit` fails when a single commit increases the size of `bin/podman` with more than 51200 bytes: https://cirrus-ci.com/task/6191166545723392?logs=main#L144
> 
> We can overwrite this with a label.

@Luap99 Would it be possible to add a label to overwrite this requirement?

TomSweeneyRedHat commented:
@Luap99 is this GTG?  PTAL

TomSweeneyRedHat commented on docs/source/markdown/podman-container-checkpoint.1.md:
```diff
@@ -28,6 +28,60 @@ archives. Not compressing the checkpoint archive can result in faster checkpoint
 
+#### **--create-image**=*image*
+
+Create a checkpoint image from running container. This is a standard OCI image
```
```suggestion
Create a checkpoint image from a running container. This is a standard OCI image
```

TomSweeneyRedHat commented on docs/source/markdown/podman-container-checkpoint.1.md:
```diff
@@ -28,6 +28,60 @@ archives. Not compressing the checkpoint archive can result in faster checkpoint
+
+Create a checkpoint image from running container. This is a standard OCI image
+created in the local image store. It consists of a single layer that contains
+all checkpoint files. The content of this image layer is in the same format as a
```
```suggestion
all of the checkpoint files. The content of this image layer is in the same format as a
```

TomSweeneyRedHat commented on docs/source/markdown/podman-container-restore.1.md:
````diff
@@ -175,6 +178,15 @@ $ podman run --rm -p 2345:80 -d webserver
 # podman container restore -p 5432:8080 --import=dump.tar
 ```
 
+Start a container with name "foobar-1". Create a checkpoint image "foobar-checkpoint". Restore the container from checkpoint image with different name.
````
```suggestion
Start a container with the name "foobar-1". Create a checkpoint image "foobar-checkpoint". Restore the container from the checkpoint image with a different name.
```

vrothberg reviewed (commented):
LGTM, thanks!

@Luap99 PTAL

rhatdan commented:
/approve

Luap99 commented:
/lgtm

Sorry for the delay.
